### PR TITLE
Added a check, to only display "execute file" action when applicable

### PR DESCRIPTION
**Pull request by lukaskremla**
I've noticed that the "Execute file in REPL" functionality doesn't have a check for whether or not the file can be executed. You could run the action when a folder was selected, when non `.py` files were selected, or on selections of multiple files. 

I've added a check to ensure that the action is only enabled if a single `.py` file is selected.

If it is not, the action will be greyed out.

Only thing that I wasn't entirely sure about is whether to modify the action further, to not show the option at all under certain circumstances, though the way the check works in this commit seems sufficient to me.

### CODE REVIEW

**Comment by lukaskremla** [comment]
In the second commit I have added similar checks to the "Upload file(s)" action. In the pre-alpha discussion we have discussed the problem of uploading directories and files at once, that if the selection contains files and containing folder, then only the first selected thing will get uploaded. Well, this makes it a bit more clear.

File: src/main/kotlin/com/jetbrains/micropython/nova/actions.kt [hunk 1]
```diff
@@ -170,8 +170,12 @@ class InstantRun : DumbAwareAction() {
     override fun getActionUpdateThread(): ActionUpdateThread = BGT
 
     override fun update(e: AnActionEvent) {
-        e.presentation.isEnabled = e.getData(CommonDataKeys.VIRTUAL_FILE) != null
-
+        val file = e.getData(CommonDataKeys.VIRTUAL_FILE)
+        val files = e.getData(CommonDataKeys.VIRTUAL_FILE_ARRAY)
+        e.presentation.isEnabled = file != null &&
+                !file.isDirectory &&
+                file.extension == "py" &&
```
Comment (elmot) [inline src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:177]: I don't see a point to prevent users from executing non-.py files.
BTW, the proper way to check the file type is `virtualFile.fileType == PythonFileType.INSTANCE`

File: src/main/kotlin/com/jetbrains/micropython/nova/actions.kt [hunk 2]
```diff
@@ -170,8 +170,12 @@ class InstantRun : DumbAwareAction() {
     override fun getActionUpdateThread(): ActionUpdateThread = BGT
 
     override fun update(e: AnActionEvent) {
-        e.presentation.isEnabled = e.getData(CommonDataKeys.VIRTUAL_FILE) != null
-
+        val file = e.getData(CommonDataKeys.VIRTUAL_FILE)
```
Comment (elmot) [inline src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:173]: There is no need to separately getting a single file and the file array.
It's better to get the array and then check if there is exactly one member, and it is a file.

File: src/main/kotlin/com/jetbrains/micropython/nova/actions.kt [hunk 3]
```diff
@@ -259,25 +263,45 @@ class OpenMpyFile : ReplAction("Open file", true) {
...
+                    return
+
+                if (file.isDirectory)
+                    directoryCount++
+                else
+                    normalFileCount++
+            }
+
+            if (normalFileCount == 1)
+                e.presentation.text = "Upload File to MicroPython Device"
```
Comment (elmot) [inline src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: 1. It's not honest to promise uploading the whole selection while we're uploading the only one file/directory.
2. Incorrect algorithm - if there are three directories and one file selected, text is going to be "Upload file..."

File: src/main/kotlin/com/jetbrains/micropython/nova/actions.kt (hunk 1 above)
Comment (lukaskremla) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:177]: > I don't see a point to prevent users from executing non-.py files. BTW, the proper way to check the file type is `virtualFile.fileType == PythonFileType.INSTANCE`

1. I suppose you are right on this, I'll change it.
2. Thanks for letting me know!

File: src/main/kotlin/com/jetbrains/micropython/nova/actions.kt (hunk 3 above)
Comment (lukaskremla) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: 1 .Could you please clarify this point? For me at least, when uploading my MicroPython project to the board, I would expect to somehow be notified if the whole selection I made won't be uploaded for some reason. Or alternatively if I see the option greyed our completely with specific selections, it will make me look at all the file tree entries I selected and experiment with different combinations until I realize what the disallowed pattern is. 

If my selection is invalid, but the plugin let's me start uploading it and then I don't see all the files I thought I was uploading, I'll think there is a bug or that something is wrong - I've already reported this behavior to you as a bug before.

2 .You're entirely right here, but I'd like to propose a different approach for the "Upload file" action. Me and my other team members agree that we rarely ever want to select a single file in our project and upload it directly to the root of the device File System. We always want to keep the structure we have created with the source root marked folders and potential folders in it ("lib" etc.).

The most common scenario for using this action for us currently is working with a single file, modifying it, and then uploading it to the device to test it as a part of the whole uploaded application instead of always flashing the entire project.

What I propose is modifying this action upload behavior to be consistent with the upload algorithm modifications I've made in #343, which keep the structure inside of the source root folders after uploading to the device. If you think that the existing functionality, where the file is uploaded directly to the device File System root, should be kept, then I say let's add a second action, make one behave the way I suggest and the other one the way things work right now.

If the single file upload action is modified in this way, then the problem of the selection not getting uploaded in its entirety in some cases is avoided.

If you'd agree with change, then it should probably be done in #343, but since this topic came up in here, I've decided to discuss it here.

Comment (elmot) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: At this moment upload action supports uploading of a single item, and its textual description reflects this.

If we decide to change the action behavior to upload whole selection, then both the description and the functionality should be changed. In this change the description is changed, but the functionality is not. Let's not do that.

In my opinion, better to implement whole selection upload, and there are problems why it's not done at the moment:
* there may be a very odd selection, how do we upload this:
image
* there should be a comprehensive set of tests
* no resources 

before merging of your #343, I need to make that testing framework, and only then it's time to improve the upload action.

My scenario of uploading a single file is when I have for instance a `boot.py` activating the wifi, and a secrets file with passwords and other stuff like mqtt credentials. That secrets file is in a separate folder, is never committed to git, and there may be several variants of it.

Your scenario better fits to "Run Configuration" approach, with some additional improvements.
For example, while the board is connected, we may keep crc or last upload timestamp of all the device files and upload only changes.
This seems to be possible, probably the overhead is not that big.
From python developer point of view, that is more reliable, it's easier to to keep all the changed files in sync with the device.

Comment (lukaskremla) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: My intent with the upload modification is to change the behavior of the single file upload action. Let's say I have a file with this path `Project/SrcMarkedFolder/Lib/logger.py`. Now when I am uploading the entire project, the resulting path on device will be `lib/logger.py`. However, if I upload using the action, then the resulting path on device will be `logger.py`.

This, in my opinion and the opinion of my colleagues does not make sense. If I wanted to to upload logger directly to the root, then I would not put it inside of the `lib` folder in the first place. What we expect the behavior to be when uploading a single file through the action is defacto a targeted version of the full project upload, which can be used when I only want to upload a single file that was modified.

But currently the behavior is as I have described already, the file always goes to the File System root without the path it "should" have, and it is my understanding that this is the intended behavior.

However, this behavior seems problematic for many reasons. First of all it is inconsistent with the project upload algorithm and for that reason it won't feel intuitive. Additionally other `.py` files in my project might be dependent upon this uploaded file in terms of imports, for example `import lib.logger`. Having the upload algorithm of this single file upload action differ from that of the project upload run action complicates the situation with imports.

Now, if the same algorithm was adapted both for whole project uploads and single file uploads, then the problem with multiple file selections goes away as well.

Is there actually some reason to make this file upload action behave the way it does now? As in it always uploads just the file without its path to the File System? All of my colleagues prefer the file upload action to be a targeted version of the project wide flashing algorithm for small changes. If I wanted the file to be uploaded directly into the device root, I would position it in my project accordingly, just like I do with my `main.py` or with my `boot.py`.

Comment (elmot) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: > Is there actually some reason to make this file upload action behave the way it does now? 
There is no reason to keep it working like now.

Your intent is clear enough, and it definitely makes sense. But there are corner cases, and the action behavior should be defined for all of them.

Case A. We have `Project/SrcMarkedFolder/Lib/logger.py`, and `Project/UnmarkedFolder/lib/logger.py` and user tries to upload the latter file. What is the destination path for the file then?

Case B. There is a project with no source folders marked, and a file `Project/Lib/logger.py` what is the destination path for the file?

Case C. We have `Project/SrcMarkedFolder/Lib/logger.py` and `Project/SrcMarkedFolder/Lib/display.py`. The user selected SrcMarkedFolder and logger.py, then clicked Upload.  Is `display.py` uploaded to 'Lib/display.py' on the device?

Case D. The user selected a bunch of files, folders and subfolders(see my screenshot) but `src` is not marked as a source root. Which files are uploaded, and what are the destination paths?

Comment (elmot) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: Well, according this particular PR - I'd extract "execution" part to another PR, and rename this one to "Upload action redesign" or something like that.

Comment (lukaskremla) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: Alright, I'll try to propose a behavior for all these scenarios.

A. Only source root marked items and under certain circumstances test root marked items get uploaded with the source/test folder itself being absent from the final on-device path. So for case A. `Project/SrcMarkedFolder/Lib/logger.py` will go on the device with this path: `Lib/logger.py` and `Project/UnmarkedFolder/lib/logger.py` will go on the device as: `UnmarkedFolder/lib/logger.py`.

B. You have outlined in other issues and discussions an algorithm to deal with this kind of uploads. If at least one folder is marked as a source root, then only source root marked folder contents will be uploaded. If no folder is marked as a source root, then files and folders in the project root will be uploaded, while excluded folders, items with a leading dot and unless explicitly being uploaded test folders are all ignored for this upload. So, `Project/Lib/logger.py` will be uploaded to the device with the following path: `Lib/logger.py`.

C. Yes, SrcMarkedFolder will always be absent from the final on-device path, so for this scenario `display.py` will be uploaded to 'Lib/display.py'.

D. In this case, paths will be created the same way as they would be in case B. Now, the simplest way to handle these odd selections in my opinion is as follows: We take the selection and if directories are present in this selection  we look for the Parent directories of the selection. Let's say there are several folders with the following paths:  "src/lib" "src/lib/wifi" "src/web"

Now we select the web, lib and src folders, and possibly files contained in them. What we do is look for the parent directories of them all, in this case we would find that src contains all of the selected items, and thus we upload all items contained in the src folder.

In practice the way I would implement this algorithm is we take all selected items and loop over them. If we find a file we just add it to a list, if we find a folder we take all of its contents and add them to the list. Now at the end of this loop we look for duplicates and remove them, now this list can be uploaded with path and upload logic we have discussed previously.

Comment (lukaskremla) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: If you'd agree with this approach, I can modify this pull request and #343 to implement it.

Comment (elmot) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: > Project/UnmarkedFolder/lib/logger.py will go on the device as: UnmarkedFolder/lib/logger.py.

That breaks my workflow, if I have files like `secrets_for_board_a/secrets.py` or `utilities-scan-i2c-for-pcb-verification/main.py` for service purposes.

> C. Yes, SrcMarkedFolder will always be absent from the final on-device path, so for this scenario display.py will be uploaded to 'Lib/display.py'.

the neighbor file is explicitly selected, and this file is explicitly left unselected, is the file uploaded at all? 

Corner case D: there is an excluded folder with some .py files inside. Is there a way to upload the folder or a file inside to the device?

Comment (lukaskremla) [reply src/main/kotlin/com/jetbrains/micropython/nova/actions.kt:287]: Well regarding your workflow, it seems easier to mark `secrets_for_board_a` as a source folder, now `secrets.py` will be uploaded to the device root and for the utilities `main.py`, you can take the same approach as you would with `secrets.py`. Alternatively, if it is just some test that needs to rarely be executed, why not rely on the "execute file in REPL" functionality?

It seems like a better approach to keep the same algorithm for full project uploads and these single file uploads. Me and me colleagues, we have all been surprised by the current behavior of this action, which is inconsistent with the whole project flashing behavior.

If you really want to still have the ability to upload an item directly to the source root like this, then I deem it more logical to create a separate action and name it as such. For example "Upload Item(s) to device root".

Sorry I have missed a part of C. Here all of SrcMarkedFolder contents will be uploaded similarly to D which I have described in a previous reply.

And for the new corner case, currently excluded folders are always excluded and ignore, this is what I have understood is supposed to happen from our previous discussions. However, test folders for example, with the new logic I propose, if you select a test folder for upload its contents will be uploaded directly to the device root, and if you select a single file, it will also be uploaded directly to the device root, though this can be discussed if you think the behavior should be different. I don't really have a set opinion on how this should be handled.

**Comment by lukaskremla** [comment]
Closing this PR as the changes have been consolidated into #347 due to being too intertwined with ones from other PRs.
